tmt/steps/provision/libvirt.py

```diff
@@ -0,0 +1,157 @@
+import re
+from pathlib import Path
+from time import sleep
+from typing import Tuple
+
+import click
+
+import tmt
+
+VM_CLONE_OUT_RE = re.compile(".+'(.+)' created successfully")
+VM_IP_RE = re.compile(".+ipv4\\s+(.+)\\/\\d+")
+VM_WAITING_COUNTER_MAX_NET = 30
+
+
+class ProvisionLibvirt(tmt.steps.provision.ProvisionPlugin):
+    """Libvirt provision plugin.
+
+    Prerequisite to have the vm already prepared via libvirt.
+    Plugin clones the existing VM and do the rest steps.
+
+    Example config:
+    ```yaml
+    provision:
+        how: libvirt
+        origin_vm_name: c2r_centos7_template
+        develop: true
+    ```
+
+    Needed packages:
+      virt-install \
+      virt-manager \
+      virt-viewer \
+      virt-viewer \
+      qemu-kvm \
+      libvirt-devel \
+      libvirt-daemon-config-network \
+      libvirt-daemon-kvm \
+      libguestfs-tools
+    """
+
+    # Guest instance
+    _guest = None
+
+    # Supported keys
+    _keys = ["origin_vm_name", "develop"]
+
+    # Supported methods
+    _methods = [tmt.steps.Method(name="libvirt", doc=__doc__, order=50)]
+
+    @classmethod
+    def options(cls, how=None):
+        """Prepare command line options for connect"""
+        return [
+            click.option(
+                "-o",
+                "--origin_vm_name",
+                help="The name of the vm to clone",
+                ),
+            click.option(
+                "--develop",
+                metavar="DEVELOP",
+                help="Do not remove the vm",
+                ),
+            ] + super().options(how)
+
+    def wake(self, options=None, data=None):
+        """Override options and wake up the guest"""
+        if data:
+            self._guest = GuestLibvirt(
+                data,
+                name=self.name,
+                parent=self.step,
+                new_keys=tuple(self._keys),
+                )
+            self._guest.prepare_vm(self.get("origin_vm_name"))
+
+    def go(self):
+        """Provision the VM."""
+        super().go()
+        # Prepare data for the guest instance
+        data = dict()
+        for key in self._keys:
+            data[key] = self.get(key)
+        self._guest = GuestLibvirt(
+            data,
+            name=self.name,
+            parent=self.step,
+            new_keys=tuple(self._keys),
+            )
+        self._guest.prepare_vm(self.get("origin_vm_name"))
+
+    def guest(self):
+        """Return the provisioned guest"""
+        return self._guest
+
+    def show(self):
+        """Show provision details"""
+        super().show(self._keys)
+
+
+class GuestLibvirt(tmt.Guest):
```

Comment:
Because the plugin needs to reference its guest class, you will need to change the order of guest and plugin classes, guest must be defined first now.